Override Dictionary save method; allow passing filename or file handle.

macks22 opened the pull request:
Seeing that the inherited `save` method includes a great deal of special cases that are irrelevant for a dictionary, the most efficient resolution is to override. I considered including the same method parameters, but none of them are really relevant. `ignore` could be used, but leaving out any dictionary attributes would likely cause errors after loading again. Resolves #291.

piskvorky commented:
Thanks. The same `file_or_handle` functionality could be useful for other classes, too. I'd prefer having it in the base class (=not override Dictionary only), it's more future-proof.

Also, use the highest protocol (-1) for pickling, to save space & memory.

macks22 commented:
I moved the alternate functionality to the `SaveLoad` base class and changed the pickle protocol to `pickle.HIGHEST_PROTOCOL`, which reads more transparently than -1. I also renamed the pickling helper functions to `smart_pickle` and `smart_unpickle` and altered the pickle import to use the name `pickle` rather than `_pickle`. This will reduce the overhead required for other developers to understand the code in the future.

piskvorky commented:
-1 on the pickle renaming. The idea is OK, but it's not backward compatible and there's tons of code that relies on `utils.pickle`. We can't just change the API like that.

The rename also seems to be the reason why all Travis tests fail.

macks22 commented:
Roger that. I reverted the pickle renaming.

piskvorky commented on gensim/utils.py:
```diff
@@ -20,7 +20,7 @@
     import cPickle as _pickle
 except ImportError:
-    import pickle as _pickle
+    import _pickle
```
Where does this `_pickle` come from? Travis fails on Python3, I think it's related.

macks22 commented:
Hmm... I guess the tests missed this case. I assumed all tests passing locally covered the bases. It's been fixed in the most recent commit. Sorry about that.

piskvorky commented:
Great, merging. Much thanks, Mack!

Re. local tests: depends which Python you run them with. If you ran them with Python3 locally, you'd see the errors. If you only run them with Python 2 (as I assume you did), you'd miss those.

That's why we use Travis, for peace for mind :)
